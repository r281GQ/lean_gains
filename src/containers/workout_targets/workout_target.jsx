import React, { PureComponent } from 'react';
import moment from 'moment';
import ImmutablePropTypes from 'react-immutable-proptypes';
import PropTypes from 'prop-types';
import { reduxForm, formValues, SubmissionError } from 'redux-form/immutable';
import { connect } from 'react-redux';
import { required } from './../../services/validators';
import { Map } from 'immutable';
import * as _ from 'lodash'

import {
  createWorkoutTarget,
  updateWorkoutTarget
} from './../../store/actionCreators/user_details_action_creators';
import WorkoutTargetForm from './../../components/workout_target/workout_target_form';

// TODO: div class wrogn on createform
class WorkoutTarget extends PureComponent {
  componentDidMount() {
    const { defaultValue, initialize } = this.props;
    if (defaultValue) {
      initialize(
        defaultValue.withMutations(map =>
          map
            .update('startDayofTraining', value => moment(value).toDate())
            .set(
              'monday',
              defaultValue.get('onDays').find(value => value === 1)
            )
            .set(
              'tuesday',
              defaultValue.get('onDays').find(value => value === 2)
            )
            .set(
              'wednesday',
              defaultValue.get('onDays').find(value => value === 3)
            )
            .set(
              'thursday',
              defaultValue.get('onDays').find(value => value === 4)
            )
            .set(
              'friday',
              defaultValue.get('onDays').find(value => value === 5)
            )
            .set(
              'saturday',
              defaultValue.get('onDays').find(value => value === 6)
            )
            .set(
              'sunday',
              defaultValue.get('onDays').find(value => value === 7)
            )
        )
      );
    } else {
      initialize(Map().set('type', 'main'));
    }
  }

  render() {
    const {
      match,
      handleSubmit,
      isCycledTraining,
      createWorkoutTarget,
      updateWorkoutTarget,
      reset
    } = this.props;
    return (
      <WorkoutTargetForm
        isCycledTraining = {isCycledTraining}
        formatDate={value => (value === '' ? 0 : value)}
        validateWorkoutName={required('Workout must have a name!')}
        submitHandler={handleSubmit(formProps => {
          const onDays = _.filter(
            [
              formProps.get('monday') ? 1 : undefined,
              formProps.get('tuesday') ? 2 : undefined,
              formProps.get('wednesday') ? 3 : undefined,
              formProps.get('thursday') ? 4 : undefined,
              formProps.get('friday') ? 5 : undefined,
              formProps.get('saturday') ? 6 : undefined,
              formProps.get('sunday') ? 7 : undefined
            ],
            item => item !== undefined
          );

          if (
            formProps.get('isCycledTraining') === 'fix' &&
            _.isEmpty(onDays)
          ) {
            throw new SubmissionError({
              isCycledTraining:
                'on fixed days you must select at least one day',
              _error: 'days'
            });
          }

          if (
            formProps.get('isCycledTraining') === 'cycle' &&
            (!formProps.get('startDayofTraining') ||
              !formProps.get('onEveryxDay'))
          )
            throw new SubmissionError({
              isCycledTraining:
                'On cycled training you must set the starting date and on which days the training should occur!',
              _error: 'exercise name'
            });

          if (
            typeof formProps.get('exercises') === 'undefined' ||
            formProps.get('exercises').isEmpty() ||
            formProps.get('exercises').every(exercise => !exercise)
          )
            throw new SubmissionError({
              isCycledTraining: 'You must provide at least one valid exercise!',
              _error: 'exercise name'
            });



          match.params.id
            ? updateWorkoutTarget({
                ...formProps.toJS(),
                startDayofTraining: moment(
                  formProps.get('startDayofTraining')
                ).valueOf(),
                onDays,
                _id: match.params.id
              })
            : createWorkoutTarget({
                ...formProps.toJS(),
                startDayofTraining: moment(
                  formProps.get('startDayofTraining')
                ).valueOf(),
                onDays: formProps.get('isCycledTraining') === 'fix' ?  onDays : []
              });

          reset();
        })}
      />
    );
  }
}


WorkoutTarget.propTypes = {
  defaultValue: ImmutablePropTypes.map,
  updateWorkoutTarget: PropTypes.func.isRequired,
  createWorkoutTarget: PropTypes.func.isRequired,
  isCycledTraining: PropTypes.oneOf(['fix' , 'cycle'])
}

export default connect(null, { createWorkoutTarget, updateWorkoutTarget })(
  reduxForm({
    form: 'workout-target'
  })(formValues('isCycledTraining')(WorkoutTarget))
);
