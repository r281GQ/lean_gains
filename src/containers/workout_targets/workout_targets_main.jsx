import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';

import {
  setSelectedWorkoutTarget,
  openWorkoutModal
} from './../../store/actionCreators/app_action_creators';
import CardList from './../../components/workout_target/card_list';
import CreateButton from './../../components/create_button';

const WorkoutTargetsMainContainer = ({
  workoutTargets,
  openWorkoutModal,
  setSelectedWorkoutTarget
}) =>
  <div>
    <CardList
      editLink="/app/workouttargets/edit/"
      collection={workoutTargets.toJS()}
      onModalStateChange={openWorkoutModal}
      setSelectedItem={setSelectedWorkoutTarget}
    />
    <CreateButton link={`/app/workouttargets/create`} />
  </div>;

const mapStateToProps = state => {
  return {
    workoutTargets: state.getIn(['userDetails', 'workoutTargets']),
    selectedWorkoutTarget: state.getIn(['app', 'selectedWorkoutTarget'])
  };
};

WorkoutTargetsMainContainer.propTypes = {
  workoutTargets: ImmutablePropTypes.map.isRequired,
  openWorkoutModal: PropTypes.func.isRequired,
  setSelectedWorkoutTarget: PropTypes.func.isRequired
};

export default connect(mapStateToProps, {
  setSelectedWorkoutTarget,
  openWorkoutModal
})(WorkoutTargetsMainContainer);

export {WorkoutTargetsMainContainer as PureWorkoutTargetsMainContainer }
