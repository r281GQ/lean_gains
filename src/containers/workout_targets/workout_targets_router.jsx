import React from 'react';
import { connect } from 'react-redux';
import { Route } from 'react-router-dom';

import WorkoutTarget from './workout_target';
import WorkoutTargetMain from './workout_targets_main';

//TODO: use HOC for modal
const WorkoutTargetsRouter = ({ workoutTargets }) =>
  <div>
    <Route
      exact
      path={'/app/workouttargets/create'}
      component={WorkoutTarget}
    />

    <Route
      exact
      path={'/app/workouttargets/edit/:id'}
      render={props => {
        const selectedWorkoutTarget = workoutTargets.find(
          (value, key) => props.match.params.id === key
        );

        return (
          <WorkoutTarget {...props} defaultValue={selectedWorkoutTarget} />
        );
      }}
    />

    <Route exact path={'/app/workouttargets'} component={WorkoutTargetMain} />
  </div>;

export default connect(state => ({
  workoutTargets: state.getIn(['userDetails', 'workoutTargets'])
}))(WorkoutTargetsRouter);
