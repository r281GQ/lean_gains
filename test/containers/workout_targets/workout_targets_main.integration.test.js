import React from 'react';
import PropTypes from 'prop-types';
import injectTapEventPlugin from 'react-tap-event-plugin';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import { mount } from 'enzyme';
import { createStore } from 'redux';
import { combineReducers } from 'redux-immutable';
import { reducer as formReducer } from 'redux-form/immutable';
import { fromJS, Map } from 'immutable';
import { StaticRouter } from 'react-router-dom';

import { PureWorkoutTargetsMainContainer } from './../../../src/containers/workout_targets/workout_targets_main';

const muiTheme = getMuiTheme();

const store = createStore(
  combineReducers({
    app: () => fromJS({ isErrorModalOpen: false, message: [''] }),
    formReducer
  })
);

injectTapEventPlugin();

describe('WorkoutTargetsMain integration test', () => {
  it('should render', () => {
    const component = mount(
      <StaticRouter location="" context={{}}>
        <PureWorkoutTargetsMainContainer
          workoutTargets={Map()}
          openWorkoutModal={() => undefined}
          setSelectedWorkoutTarget={() => undefined}
        />
      </StaticRouter>,
      {
        context: { muiTheme, store },
        childContextTypes: {
          muiTheme: PropTypes.object,
          store: PropTypes.object
        }
      }
    );

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });
});
