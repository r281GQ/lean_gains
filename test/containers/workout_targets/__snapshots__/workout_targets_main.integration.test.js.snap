// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkoutTargetsMain integration test should render 1`] = `
<StaticRouter
  basename=""
  context={Object {}}
  location=""
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <WorkoutTargetsMainContainer
      openWorkoutModal={[Function]}
      setSelectedWorkoutTarget={[Function]}
      workoutTargets={
        Immutable.Map {
        }
      }
    >
      <div>
        <CardList
          collection={Object {}}
          editLink="/app/workouttargets/edit/"
          onModalStateChange={[Function]}
          setSelectedItem={[Function]}
        >
          <List>
            <div
              style={
                Object {
                  "muiPrepared": true,
                  "padding": "8px 0px 8px 0px",
                }
              }
            />
          </List>
        </CardList>
        <CreateButton
          link="/app/workouttargets/create"
        >
          <Link
            onClick={[Function]}
            replace={false}
            to="/app/workouttargets/create"
          >
            <a
              href="/app/workouttargets/create"
              onClick={[Function]}
            >
              <FloatingActionButton
                className="fixed-floating-button"
                disabled={false}
                mini={false}
                secondary={false}
                zDepth={2}
              >
                <Paper
                  circle={true}
                  className="fixed-floating-button"
                  rounded={true}
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "display": "inline-block",
                      "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                    }
                  }
                  transitionEnabled={true}
                  zDepth={2}
                >
                  <div
                    className="fixed-floating-button"
                    style={
                      Object {
                        "WebkitTapHighlightColor": "rgba(0,0,0,0)",
                        "backgroundColor": "transparent",
                        "borderRadius": "50%",
                        "boxShadow": "0 3px 10px rgba(0, 0, 0, 0.16),
                               0 3px 10px rgba(0, 0, 0, 0.23)",
                        "boxSizing": "border-box",
                        "color": "rgba(0, 0, 0, 0.87)",
                        "display": "inline-block",
                        "fontFamily": "Roboto, sans-serif",
                        "muiPrepared": true,
                        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                      }
                    }
                  >
                    <EnhancedButton
                      containerElement="button"
                      disabled={false}
                      focusRippleColor="#ffffff"
                      onBlur={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onKeyboardFocus={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchStart={[Function]}
                      onTouchTap={[Function]}
                      style={
                        Object {
                          "backgroundColor": "#00bcd4",
                          "borderRadius": "50%",
                          "height": 56,
                          "overflow": "hidden",
                          "padding": 0,
                          "textAlign": "center",
                          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                          "verticalAlign": "bottom",
                          "width": 56,
                        }
                      }
                      tabIndex={0}
                      touchRippleColor="#ffffff"
                      type="button"
                    >
                      <button
                        disabled={false}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchStart={[Function]}
                        onTouchTap={[Function]}
                        style={
                          Object {
                            "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
                            "backgroundColor": "#00bcd4",
                            "border": 10,
                            "borderRadius": "50%",
                            "boxSizing": "border-box",
                            "cursor": "pointer",
                            "display": "inline-block",
                            "fontFamily": "Roboto, sans-serif",
                            "fontSize": "inherit",
                            "fontWeight": "inherit",
                            "height": 56,
                            "margin": 0,
                            "muiPrepared": true,
                            "outline": "none",
                            "overflow": "hidden",
                            "padding": 0,
                            "position": "relative",
                            "textAlign": "center",
                            "textDecoration": "none",
                            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                            "verticalAlign": "bottom",
                            "width": 56,
                            "zIndex": 1,
                          }
                        }
                        tabIndex={0}
                        type="button"
                      >
                        <TouchRipple
                          abortOnScroll={true}
                          color="#ffffff"
                        >
                          <div
                            onMouseDown={[Function]}
                            onMouseLeave={[Function]}
                            onMouseUp={[Function]}
                            onTouchEnd={[Function]}
                            onTouchStart={[Function]}
                          >
                            <div
                              style={
                                Object {
                                  "muiPrepared": true,
                                  "top": 0,
                                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                                }
                              }
                            >
                              <ContentAdd
                                style={
                                  Object {
                                    "color": "#ffffff",
                                    "fill": "#ffffff",
                                    "height": 56,
                                    "lineHeight": "56px",
                                  }
                                }
                              >
                                <SvgIcon
                                  onMouseEnter={[Function]}
                                  onMouseLeave={[Function]}
                                  style={
                                    Object {
                                      "color": "#ffffff",
                                      "fill": "#ffffff",
                                      "height": 56,
                                      "lineHeight": "56px",
                                    }
                                  }
                                  viewBox="0 0 24 24"
                                >
                                  <svg
                                    onMouseEnter={[Function]}
                                    onMouseLeave={[Function]}
                                    style={
                                      Object {
                                        "color": "#ffffff",
                                        "display": "inline-block",
                                        "fill": "#ffffff",
                                        "height": 56,
                                        "lineHeight": "56px",
                                        "muiPrepared": true,
                                        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                                        "userSelect": "none",
                                        "width": 24,
                                      }
                                    }
                                    viewBox="0 0 24 24"
                                  >
                                    <path
                                      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                                    />
                                  </svg>
                                </SvgIcon>
                              </ContentAdd>
                            </div>
                          </div>
                        </TouchRipple>
                      </button>
                    </EnhancedButton>
                  </div>
                </Paper>
              </FloatingActionButton>
            </a>
          </Link>
        </CreateButton>
      </div>
    </WorkoutTargetsMainContainer>
  </Router>
</StaticRouter>
`;
